Support multiple HTTP headers with same name

dinkelacker opened the pull request:
This is a draft. It is currently missing the implementation for macOS resp. URLSession.

Closes #24878.

dinkelacker committed:
Add GetAllHeaderValues() to wxWebResponse

HTTP headers like Set-Cookie can be present multiple times in a
web response by definition. The existing GetHeader() method is
not sufficient to cope with this situation.

This commit already adds an implementation for Windows.
Linux and macOS follow in the next commits.

dinkelacker committed:
Add support for multiple headers with same name to wxWebRequest

wxWebRequest::AddHeader() was added to accompany the existing
SetHeader() method. In wxWebSession, AddCommonHeader() did exist
already but behaved like setting a header. Its behavior was adapted
to its name and SetCommonHeader() was added with the old behavior
to accompany AddCommonHeader().

The WinHTTP and CURL implementations of wxWebRequest do now append
multiple headers of the same name. The URLSession implementation
follows in the next commits.

vadz commented on src/msw/webrequest_winhttp.cpp:
```diff
@@ -214,6 +214,38 @@ static wxString wxWinHTTPQueryHeaderString(HINTERNET hRequest, DWORD dwInfoLevel
     return result;
 }
 
+static std::vector<wxString> wxWinHTTPQueryAllHeaderStrings(HINTERNET hRequest, DWORD dwInfoLevel,
```
Could this be refactored to reuse the existing `wxWinHTTPQueryHeaderString()` by adding an optional `DWORD* nextIndex = WINHTTP_NO_HEADER_INDEX` parameter to it and passing it `&nextIndex` when calling it from here?

The code dealing with buffers is sufficiently non trivial for me not to want to duplicate it...

vadz commented on src/common/webrequest.cpp:
```diff
@@ -702,10 +709,10 @@ wxString wxWebResponseImpl::GetSuggestedFileName() const
     wxWebRequestHeaderMap params;
     const wxString disp = wxPrivate::SplitParameters(contentDisp, params);
-    if ( disp == "attachment" )
+    if ( disp == "attachment" && !params["filename"].empty())
```
Exceedingly nitpicky, but:
```suggestion
    if ( disp == "attachment" && !params["filename"].empty() )
```

vadz commented on src/osx/webrequest_urlsession.mm:
```diff
@@ -434,6 +434,12 @@ void Wait()
+std::vector<wxString> wxWebResponseURLSession::GetAllHeaderValues(const wxString& WXUNUSED(name)) const
+{
+    std::vector<wxString> result;
+    return result;
```
It's fine not to implement this for macOS if you don't know how to do it but I'd at least call `GetHeader()` here as returning the first header value seems like a more useful fallback than not returning anything at all.

This behaviour should be mentioned in the documentation, of course.

vadz reviewed (commented):
Thanks for this!

The API looks good to me and a separate thank you for updating the documentation too!

Once the CI builds are fixed, this can be merged.

dinkelacker committed:
Generalize CheckExpectedJSON()

dinkelacker committed:
Add headers test

The contents of a webrequest header can be checked with the
headers endpoint of the httpbin service.

dinkelacker committed:
Add fallback to URLSession webrequest backend

Multiple HTTP headers with the same name are not supported by the
URLsession API. As a fallback, the last header is used instead.

dinkelacker committed:
Improve test output on failure

The two strings that are compared should now be printed on
test failure.

dinkelacker committed:
Refactor wxWinHTTPQueryHeaderString()

Some code duplication can be removed by using
wxWinHTTPQueryAllHeaderStrings() in wxWinHTTPQueryHeaderString()
and returning ony the last queried header. This change also
increases consistency accross all webrequest implementations to
prefer the last header over the first in ambiguities.

dinkelacker committed:
Fix newly introduced headers test

The format of the generated JSON by the httpbin service for
multiple headers with the same name differs between platforms.
These headers are sometimes put into a JSON array and sometimes
concatenated into a single comma-separated string.

dinkelacker committed:
Fix queries of non-existent HTTP header strings

vadz commented on src/osx/webrequest_urlsession.mm:
```diff
@@ -190,8 +190,10 @@ - (void)URLSession:(NSURLSession *)session
     {
-        [req setValue:wxCFStringRef(it->second).AsNSString() forHTTPHeaderField:
-         wxCFStringRef(it->first).AsNSString()];
+        // TODO: URLSession does not support multiple headers with the same name.
```
FWIW there apparently is `[NSURLSession addValue:forHTTPHeaderField:]`, so you could try using it here if you're motivated.

vadz commented on src/osx/webrequest_urlsession.mm:
```diff
@@ -434,6 +436,13 @@ void Wait()
 
+std::vector<wxString> wxWebResponseURLSession::GetAllHeaderValues(const wxString& name) const
+{
+    // TODO: URLSession does not support multiple headers with the same name.
```
Considering `addValue` description, it would be logical — but I didn't test it at all — if `NSURLSession` also combined multiple header values into a single comma-separated string.

vadz reviewed (approved):
Thanks, we can merge this already or I can wait if you'd like to test `addValue` — please let me know if you would.

Thanks again!

vadz commented:
Sorry, I've just found a big problem with this PR: the previously working code which called

```cpp
    m_session.AddCommonHeader("User-Agent", "Whatever");
```

doesn't work any more because we end up with 2 `User-Agent:` lines in the headers passed to `WinHttpSendRequest()` due to the default "User-Agent" already present there and it just fails immediately when this happens.

This is a silent regression and it really shouldn't work like this, so I'm going to do something about it, but I'm not sure what exactly yet. Making `AddCommonHeader()` behave as `SetHeader()`, rather than as `AddHeader()`, would be very inconsistent. I guess we're going to need to avoid calling it by default in the first place.

vadz commented:
Moreover, I start to really question if we should be ever sending multiple lines for the same header in the first place. It looks like the only valid uses for multiple headers, like `Accept:`, can be also achieved by combining all their values into a single string using commas and this is something that [WinHttpAddRequestHeaders()](https://learn.microsoft.com/en-us/windows/win32/api/winhttp/nf-winhttp-winhttpaddrequestheaders) can do and while it can also do other things, it never adds multiple header lines.

vadz commented:
Sorry, after re-checking HTTP standard, I think it was a bad idea to merge this PR. It [clearly says](https://datatracker.ietf.org/doc/html/rfc9110#name-field-order)

> This means that, aside from the well-known exception noted below, a sender MUST NOT generate multiple field lines with the same name in a message

where the "well-known exception" is `Set-Cookie`.

I'm going to revert this and if there is any interest, we should add a special function specifically for cookies, but it doesn't make sense to allow sending multiple headers when the RFC explicitly says that we MUST NOT be doing this.
